### PR TITLE
Update Rust toolchain to 1.83.0

### PR DESCRIPTION
**Pull request by SimonSapin**
We had been stuck on 1.76.0 for a long time because of an unexplained panic in `rusty_v8` that only happened in release mode, but that is no longer relevant since https://github.com/apollographql/router/pull/6418!

The project compiles and passes test with no change in the new toolchain, but but rustc and clippy emit new warnings. Fixing those is most of the PR. Part of it was done by `cargo clippy --fix` to automatically apply the compiler’s suggested changes.

I have removed dead/unused code on the basis that if we want it later we can recover it from git history. This includes:

* Half of `apollo-federation/src/utils/fallible_iterator.rs`
* `SpecDefinition::minimum_federation_version`

---

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [ ] Changes are compatible[^1]
- [ ] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [ ] Unit Tests
    - [ ] Integration Tests
    - [ ] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

### CODE REVIEW

File: apollo-router/src/services/subgraph.rs [hunk 1]
```diff
@@ -238,6 +238,7 @@ impl Response {
     /// The parameters are not optional, because in a live situation all of these properties must be
     /// set and be correct to create a Response.
     #[builder(visibility = "pub")]
+    #[allow(clippy::too_many_arguments)] // not typically used directly, only defines the builder
```
Comment (SimonSapin) [inline apollo-router/src/services/subgraph.rs:241]: Perhaps `buildstructor` could generate this

**Review by BrynCooke — APPROVED** [review]
I'll fix the perf tests

File: apollo-federation/src/link/context_spec_definition.rs [hunk 2]
```diff
@@ -61,18 +59,14 @@ impl SpecDefinition for ContextSpecDefinition {
...
-    }
 }
 
 lazy_static! {
     pub(crate) static ref CONTEXT_VERSIONS: SpecDefinitions<ContextSpecDefinition> = {
         let mut definitions = SpecDefinitions::new(Identity::context_identity());
         definitions.add(ContextSpecDefinition::new(
             Version { major: 0, minor: 1 },
-            Some(Version { major: 2, minor: 8 }),
+            // minimum_federation_version = Some(Version { major: 2, minor: 8 }),
```
Comment (bnjjj) [inline apollo-federation/src/link/context_spec_definition.rs:69]:
```suggestion
```

File: apollo-federation/src/link/inaccessible_spec_definition.rs [hunk 3]
```diff
@@ -97,22 +95,18 @@ impl SpecDefinition for InaccessibleSpecDefinition {
...
-    }
 }
 
 lazy_static! {
     pub(crate) static ref INACCESSIBLE_VERSIONS: SpecDefinitions<InaccessibleSpecDefinition> = {
         let mut definitions = SpecDefinitions::new(Identity::inaccessible_identity());
         definitions.add(InaccessibleSpecDefinition::new(
             Version { major: 0, minor: 1 },
-            None,
+            // minimum_federation_version = None,
```
Comment (bnjjj) [inline apollo-federation/src/link/inaccessible_spec_definition.rs:105]:
```suggestion
```

File: apollo-federation/src/link/inaccessible_spec_definition.rs [hunk 4]
```diff
@@ -97,22 +95,18 @@ impl SpecDefinition for InaccessibleSpecDefinition {
...
         ));
         definitions.add(InaccessibleSpecDefinition::new(
             Version { major: 0, minor: 2 },
-            Some(Version { major: 2, minor: 0 }),
+            // minimum_federation_version = Some(Version { major: 2, minor: 0 }),
```
Comment (bnjjj) [inline apollo-federation/src/link/inaccessible_spec_definition.rs:109]:
```suggestion
```

File: apollo-federation/src/link/join_spec_definition.rs [hunk 5]
```diff
@@ -410,34 +408,30 @@ impl SpecDefinition for JoinSpecDefinition {
...
-    }
 }
 
 lazy_static! {
     pub(crate) static ref JOIN_VERSIONS: SpecDefinitions<JoinSpecDefinition> = {
         let mut definitions = SpecDefinitions::new(Identity::join_identity());
         definitions.add(JoinSpecDefinition::new(
             Version { major: 0, minor: 1 },
-            None,
+            // minimum_federation_version = None,
```
Comment (bnjjj) [inline apollo-federation/src/link/join_spec_definition.rs:418]:
```suggestion
```

File: apollo-federation/src/link/join_spec_definition.rs [hunk 6]
```diff
@@ -410,34 +408,30 @@ impl SpecDefinition for JoinSpecDefinition {
...
         ));
         definitions.add(JoinSpecDefinition::new(
             Version { major: 0, minor: 2 },
-            None,
+            // minimum_federation_version = None,
```
Comment (bnjjj) [inline apollo-federation/src/link/join_spec_definition.rs:422]:
```suggestion
```

File: apollo-federation/src/link/join_spec_definition.rs [hunk 7]
```diff
@@ -410,34 +408,30 @@ impl SpecDefinition for JoinSpecDefinition {
...
         ));
         definitions.add(JoinSpecDefinition::new(
             Version { major: 0, minor: 3 },
-            Some(Version { major: 2, minor: 0 }),
+            // minimum_federation_version = Some(Version { major: 2, minor: 0 }),
```
Comment (bnjjj) [inline apollo-federation/src/link/join_spec_definition.rs:426]:
```suggestion
```

File: apollo-federation/src/link/join_spec_definition.rs [hunk 8]
```diff
@@ -410,34 +408,30 @@ impl SpecDefinition for JoinSpecDefinition {
...
         ));
         definitions.add(JoinSpecDefinition::new(
             Version { major: 0, minor: 4 },
-            Some(Version { major: 2, minor: 7 }),
+            // minimum_federation_version = Some(Version { major: 2, minor: 7 }),
```
Comment (bnjjj) [inline apollo-federation/src/link/join_spec_definition.rs:430]:
```suggestion
```

File: apollo-federation/src/link/join_spec_definition.rs [hunk 9]
```diff
@@ -410,34 +408,30 @@ impl SpecDefinition for JoinSpecDefinition {
...
         ));
         definitions.add(JoinSpecDefinition::new(
             Version { major: 0, minor: 5 },
-            Some(Version { major: 2, minor: 8 }),
+            // minimum_federation_version = Some(Version { major: 2, minor: 8 }),
```
Comment (bnjjj) [inline apollo-federation/src/link/join_spec_definition.rs:434]:
```suggestion
```

File: apollo-federation/src/link/link_spec_definition.rs [hunk 10]
```diff
@@ -28,23 +22,19 @@ impl SpecDefinition for LinkSpecDefinition {
...
-    }
 }
 
 lazy_static! {
     pub(crate) static ref CORE_VERSIONS: SpecDefinitions<LinkSpecDefinition> = {
         let mut definitions = SpecDefinitions::new(Identity::core_identity());
         definitions.add(LinkSpecDefinition::new(
             Version { major: 0, minor: 1 },
-            None,
+            // minimum_federation_version = None,
```
Comment (bnjjj) [inline apollo-federation/src/link/link_spec_definition.rs:32]:
```suggestion
```

File: apollo-federation/src/link/link_spec_definition.rs [hunk 11]
```diff
@@ -28,23 +22,19 @@ impl SpecDefinition for LinkSpecDefinition {
...
             Identity::core_identity(),
         ));
         definitions.add(LinkSpecDefinition::new(
             Version { major: 0, minor: 2 },
-            Some(Version { major: 2, minor: 0 }),
+            // minimum_federation_version = Some(Version { major: 2, minor: 0 }),
```
Comment (bnjjj) [inline apollo-federation/src/link/link_spec_definition.rs:37]:
```suggestion
```

File: apollo-federation/src/link/link_spec_definition.rs [hunk 12]
```diff
@@ -53,7 +43,7 @@ lazy_static! {
         let mut definitions = SpecDefinitions::new(Identity::link_identity());
         definitions.add(LinkSpecDefinition::new(
             Version { major: 1, minor: 0 },
-            Some(Version { major: 2, minor: 0 }),
+            // minimum_federation_version = Some(Version { major: 2, minor: 0 }),
```
Comment (bnjjj) [inline apollo-federation/src/link/link_spec_definition.rs:46]:
```suggestion
```

**Review by bnjjj — DISMISSED** [review]
I can't comment on all commented lines but these need to be removed I think

File: apollo-federation/src/link/inaccessible_spec_definition.rs [hunk 13]
```diff
@@ -97,22 +95,18 @@ impl SpecDefinition for InaccessibleSpecDefinition {
     fn url(&self) -> &Url {
         &self.url
     }
-
-    fn minimum_federation_version(&self) -> Option<&Version> {
-        self.minimum_federation_version.as_ref()
-    }
 }
 
 lazy_static! {
```
Comment (TylerBloom) [inline apollo-federation/src/link/inaccessible_spec_definition.rs:100]: We should remove `lazy_static!` in favor of `std::sync::LazyLock`

**Review by TylerBloom — CHANGES_REQUESTED** [review]
We should remove `lazy_static` as a dependency from `apollo-federation`.

**Review by TylerBloom — APPROVED** [review]
The changes in `apollo-federation` look good to me.

**Pull request merged**